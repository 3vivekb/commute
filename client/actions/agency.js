import makeGenericModelActions from '../utils/actions'

const actions = makeGenericModelActions({
  commands: {
    'Collection GET': {},
    'Collection POST': {},
    'DELETE': {
      redirectionStrategy: 'toHome'
    },
    'GET': {},
    'PUT': {}
  },
  pluralName: 'agencies',
  singularName: 'agency'
})

export default actions
