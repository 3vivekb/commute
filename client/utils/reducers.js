import update from 'react-addons-update'

import {addToEntityMap, addEntitiesToEntityMap, deleteFromEntityMap, deleteManyFromEntityMap} from './entities'

function addChildEntityToParent (parentMap, newEntity, parentIdField, childPluralName) {
  return update(parentMap, {
    [newEntity[parentIdField]]: {
      [childPluralName]: {
        $push: [newEntity._id]
      }
    }
  })
}

/**
 * Make handlers for children creation and deletion events
 *
 * @param  {Object} cfg  An configuration object structured as follows:
 *   - childPluralName (String): Plural name of child (parent's child id array)
 *   - childSingularName (String): Singular name of child (for making reducer title)
 *   - parentIdField (String): Id field of parent in child entity
 * @return {Object}     Reducers
 */
export function makeChildrenHandlers (cfg) {
  const reducers = {}
  reducers[`add ${cfg.childSingularName}`] = function (state, action) {
    const newEntity = action.payload
    return addChildEntityToParent(
      state,
      newEntity,
      cfg.parentIdField,
      cfg.childPluralName
    )
  }

  reducers[`add many ${cfg.childPluralName}`] = function (state, action) {
    const newEntities = action.payload
    let updatedState = state
    newEntities.forEach(newEntity => {
      updatedState = addChildEntityToParent(
        updatedState,
        newEntity,
        cfg.parentIdField,
        cfg.childPluralName
      )
    })
    return updatedState
  }

  reducers[`delete ${cfg.childSingularName}`] = function (state, action) {
    const entity = action.payload
    const childIds = state[entity[cfg.parentIdField]][cfg.childPluralName]
    return update(state, {
      [entity[cfg.parentIdField]]: {
        [cfg.childPluralName]: {
          $set: childIds.filter((id) => id !== entity._id)
        }
      }
    })
  }

  return reducers
}

/**
 * Make generic reducers for an entity type
 *
 * @param  {Object} cfg  An object with name and handlers parameters structured as follows:
 *   - name: must have singular and plural keys with the corresponding strings
 *   - handlers: An array of strings of handlers.  Possible values: ['add', 'add many', 'delete', 'set', 'set many']
 * @return {Object}      The generated reducers
 */
export function makeGenericReducerHandlers (cfg) {
  const reducers = {
    'log out' (state, action) {
      return {}
    }
  }
  if (cfg.handlers.indexOf('add') !== -1) {
    reducers[`add ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('add many') !== -1) {
    reducers[`add many ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('delete') !== -1) {
    reducers[`delete ${cfg.name.singular}`] = function (state, action) {
      return deleteFromEntityMap(state, action.payload._id)
    }
  }

  if (cfg.handlers.indexOf('delete many') !== -1) {
    reducers[`delete many ${cfg.name.plural}`] = function (state, action) {
      return deleteManyFromEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set') !== -1) {
    reducers[`set ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set many') !== -1) {
    reducers[`set ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  return reducers
}
