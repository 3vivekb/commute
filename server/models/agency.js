const {Schema} = require('mongoose')

const Organization = require('./organization')
const trashPlugin = require('./plugins/trash')

const schema = new Schema({
  name: {
    required: true,
    type: String
  }
})

schema.pre('save', function (next) {
  // CASCADE DELETE if needed
  if (this.isModified('trashed') && this.trashed) {
    Organization.update({ agencyId: this._id }, { trashed: new Date() }).exec()
  }
  next()
})

schema.plugin(trashPlugin)

module.exports = schema
