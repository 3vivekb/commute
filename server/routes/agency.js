const models = require('../models')
const makeRestEndpoints = require('../utils/restEndpoints')

module.exports = function makeRoutes (app) {
  makeRestEndpoints(app,
    {
      childModels: [{
        foreignKey: 'agencyId',
        key: 'organizations',
        model: models.Organization
      }],
      commands: {
        'Collection GET': {},
        'Collection POST': {},
        'GET': {},
        'DELETE': {},
        'PUT': {}
      },
      model: models.Agency,
      name: 'agency'
    }
  )
}
