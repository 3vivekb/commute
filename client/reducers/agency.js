import {makeGenericReducerHandlers} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'delete', 'set', 'set all'],
  name: {
    singular: 'agency',
    plural: 'agencies'
  }
})

export const initialState = {}
