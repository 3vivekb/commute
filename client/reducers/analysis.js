import update from 'react-addons-update'

import analysisDefaults from '../utils/analysisDefaults'
import {fixedRound} from '../utils/common'
import {addToEntityMap} from '../utils/entities'
import {makeGenericReducerHandlers} from '../utils/reducers'

const METRICS = Object.keys(analysisDefaults.metrics)

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'delete', 'set many'],
  name: {
    singular: 'analysis',
    plural: 'analyses'
  }
})

reducers['receive mock calculated trips'] = function (state, action) {
  const {analysisId, trips} = action.payload
  const affectedAnalysis = state[analysisId]
  let updatedAnalysis = update(affectedAnalysis, {
    trips: { $set: trips }
  })

  // calculate array of values for each metric
  const vals = {}
  const modes = Object.keys(trips[0]).filter((mode) => (['commuterId']).indexOf(mode) === -1)
  modes.forEach((mode) => {
    vals[mode] = {
      cost: [],
      distance: [],
      time: []
    }
  })
  trips.forEach((trip) =>
    modes.forEach((mode) =>
      METRICS.forEach((metric) => {
        vals[mode][metric].push(trip[mode][metric])
      })
    )
  )

  // sort arrays
  modes.forEach((mode) =>
    METRICS.forEach((metric) => {
      vals[mode][metric].sort()
    })
  )

  updatedAnalysis = update(updatedAnalysis, {
    tripVals: { $set: vals }
  })

  // calculate analysis summary metrics
  // Potential Savings is an overall average of the difference in cost between
  // the most likely trip and the driving trip
  const len = trips.length
  let totalCostDifference = 0
  let totalTravelTime = 0
  let totalDistance = 0
  trips.forEach((trip) => {
    totalCostDifference += trip.car.cost - trip.mostLikely.cost
    totalTravelTime += trip.mostLikely.time
    totalDistance += trip.mostLikely.distance
  })

  updatedAnalysis = update(updatedAnalysis, {
    summary: {
      $set: {
        avgTravelTime: fixedRound(totalTravelTime / len),
        avgDistance: fixedRound(totalDistance / len),
        savingsPerTrip: fixedRound(totalCostDifference / len),
        savingsPerTripYear: fixedRound(totalCostDifference / len * 260),
        savingsTotalPerDay: fixedRound(totalCostDifference),
        savingsTotalPerYear: fixedRound(totalCostDifference * 260)
      }
    }
  })

  return addToEntityMap(state, updatedAnalysis)
}

export const initialState = {}
