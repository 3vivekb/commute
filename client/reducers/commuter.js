import {addToEntityMap} from '../utils/entities'
import {
  getGenericReducerInitialState,
  makeGenericReducerHandlers,
  updateTime
} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'add many', 'delete', 'set', 'set many'],
  name: {
    singular: 'commuter',
    plural: 'commuters'
  }
})

reducers['add group'] = function (state, action) {
  const newGroup = {...action.payload}
  let updatedState = {...state}
  newGroup.commuters.map((commuter) => {
    updatedState = addToEntityMap(updatedState, commuter)
  })
  return updateTime(updatedState)
}

reducers['append commuters'] = function (state, action) {
  let updatedState = state
  action.payload.commuters.forEach((commuter) => {
    updatedState = addToEntityMap(updatedState, commuter)
  })
  return updateTime(updatedState)
}

export const initialState = getGenericReducerInitialState()
