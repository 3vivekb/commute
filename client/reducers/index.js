import {handleActions} from 'redux-actions'

import * as agency from './agency'
import * as analysis from './analysis'
import * as commuter from './commuter'
import * as group from './group'
import * as organization from './organization'
import * as site from './site'
import * as user from './user'

export default {
  agency: handleActions(agency.reducers, agency.initialState),
  analysis: handleActions(analysis.reducers, analysis.initialState),
  commuter: handleActions(commuter.reducers, commuter.initialState),
  group: handleActions(group.reducers, group.initialState),
  organization: handleActions(organization.reducers, organization.initialState),
  site: handleActions(site.reducers, site.initialState),
  user: handleActions(user.reducers, user.initialState)
}
