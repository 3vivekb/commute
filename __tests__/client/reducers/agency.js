/* globals describe */

import {agencyWithAnOrganization, blankAgency, mockStores} from '../../test-utils/mock-data'
import {makeGenericReducerTestCases} from '../../test-utils/reducers'

import * as agency from '../../../client/reducers/agency'

describe('client > reducers > agency', () => {
  makeGenericReducerTestCases({
    handlers: {
      add: {
        initialState: agency.initialState,
        payload: blankAgency
      },
      delete: {
        initialState: mockStores.withBlankAgency.agency,
        payload: 'agency-1'
      },
      set: {
        initialState: mockStores.withBlankAgency.agency,
        payload: agencyWithAnOrganization
      },
      'set all': {
        initialState: agency.initialState,
        payload: [blankAgency, agencyWithAnOrganization]
      }
    },
    initialState: agency.initialState,
    name: {
      plural: 'agencies',
      singular: 'agency'
    },
    reducers: agency.reducers
  })
})
