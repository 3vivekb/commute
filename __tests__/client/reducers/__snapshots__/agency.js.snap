exports[`client > reducers > agency generic reducer tests should handle add agency 1`] = `
Object {
  "agency-1": Object {
    "_id": "agency-1",
    "name": "Mock Agency",
    "organizations": Array [],
  },
}
`;

exports[`client > reducers > agency generic reducer tests should handle delete agency 1`] = `Object {}`;

exports[`client > reducers > agency generic reducer tests should handle set agencies 1`] = `
Object {
  "agency-1": Object {
    "_id": "agency-1",
    "name": "Mock Agency",
    "organizations": Array [],
  },
  "agency-2": Object {
    "_id": "agency-2",
    "name": "Mock Agency",
    "organizations": Array [
      "organization-2",
    ],
  },
}
`;

exports[`client > reducers > agency generic reducer tests should handle set agency 1`] = `
Object {
  "agency-1": Object {
    "_id": "agency-1",
    "name": "Mock Agency",
    "organizations": Array [],
  },
  "agency-2": Object {
    "_id": "agency-2",
    "name": "Mock Agency",
    "organizations": Array [
      "organization-2",
    ],
  },
}
`;

exports[`client > reducers > agency organizations children handler reducer tests it should handle add organization 1`] = `
Object {
  "agency-1": Object {
    "_id": "agency-1",
    "name": "Mock Agency",
    "organizations": Array [
      "organization-new",
    ],
  },
}
`;

exports[`client > reducers > agency organizations children handler reducer tests it should handle delete organization 1`] = `
Object {
  "agency-3": Object {
    "_id": "agency-3",
    "name": "An agency",
    "organizations": Array [],
  },
}
`;
